File: 680-valid-palindrome-ii/680-valid-palindrome-ii.cpp
class Solution {
public:
    bool validPalindrome(string s) {
        return solve(s);
    }
private:
    bool solve(string s) {
        int i = 0, j = s.length() - 1;
        
        while(i < j) {
            if(s[i] != s[j])
                return check(s, i + 1, j) || check(s, i, j - 1);
            i++;
            --j;
        }
        return true;
    }
    
    bool check(string s, int i, int j) {
        
        while(i < j) {
            if(s[i] != s[j])
                return false;
            ++i;
            --j;
        }
        return true;
    }
};

File: 680-valid-palindrome-ii/680-valid-palindrome-ii.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        
        def check(s, i, j):
            
            while i < j:
                if s[i] != s[j]:
                    return False
                i += 1
                j -= 1
            return True
        
        i, j = 0, len(s) - 1
        
        while i < j:
            if s[i] != s[j]:
                return check(s, i + 1, j) or check(s, i, j - 1)
            i += 1
            j -= 1
            
        return True